pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3InclusionIndexFilterOperator.java

```diff
@@ -0,0 +1,145 @@
+import org.apache.pinot.core.operator.dociditerators.ScanBasedDocIdIterator;
+import org.apache.pinot.core.operator.docidsets.BitmapDocIdSet;
+import org.apache.pinot.segment.local.utils.GeometrySerializer;
+import org.apache.pinot.segment.local.utils.H3Utils;
+import org.apache.pinot.segment.spi.IndexSegment;
+import org.apache.pinot.segment.spi.index.reader.H3IndexReader;
+import org.apache.pinot.spi.utils.BooleanUtils;
+import org.apache.pinot.spi.utils.BytesUtils;
+import org.locationtech.jts.geom.Geometry;
+import org.roaringbitmap.buffer.BufferFastAggregation;
+import org.roaringbitmap.buffer.ImmutableRoaringBitmap;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+
+/**
+ * A filter operator that uses H3 index for geospatial data inclusion
+ */
+public class H3InclusionIndexFilterOperator extends BaseFilterOperator {
+
+  private static final String EXPLAIN_NAME = "INCLUSION_FILTER_H3_INDEX";
+
+  private final IndexSegment _segment;
+  private final Predicate _predicate;
+  private final int _numDocs;
+  private final H3IndexReader _h3IndexReader;
+  private final Geometry _geometry;
+  private final boolean _isPositiveCheck;
+
+  public H3InclusionIndexFilterOperator(IndexSegment segment, Predicate predicate, int numDocs) {
+    _segment = segment;
+    _predicate = predicate;
+    _numDocs = numDocs;
+
+    List<ExpressionContext> arguments = predicate.getLhs().getFunction().getArguments();
+    EqPredicate eqPredicate = (EqPredicate) predicate;
+    _isPositiveCheck = BooleanUtils.toBoolean(eqPredicate.getValue());
+
+    if (arguments.get(0).getType() == ExpressionContext.Type.IDENTIFIER) {
+      _h3IndexReader = segment.getDataSource(arguments.get(0).getIdentifier()).getH3Index();
+      _geometry = GeometrySerializer.deserialize(BytesUtils.toBytes(arguments.get(1).getLiteral()));
+    } else {
+      _h3IndexReader = segment.getDataSource(arguments.get(1).getIdentifier()).getH3Index();
+      _geometry = GeometrySerializer.deserialize(BytesUtils.toBytes(arguments.get(0).getLiteral()));
+    }
+    // must be some h3 index
+    assert _h3IndexReader != null : "the column must have H3 index setup.";
+  }
+
+  @Override
+  protected FilterBlock getNextBlock() {
+    // get the set of H3 cells at the specified resolution which completely cover the input shape and potential cover.
+    final Pair<LongSet, LongSet> fullCoverAndPotentialCoverCells =
+        H3Utils.coverGeometryInH3(_geometry, _h3IndexReader.getH3IndexResolution().getLowestResolution());
+    final LongSet fullyCoverH3Cells = fullCoverAndPotentialCoverCells.getLeft();
+    final LongSet potentialCoverH3Cells = fullCoverAndPotentialCoverCells.getRight();
+
+    // have list of h3 cell ids for polygon provided
+    // return filtered num_docs
+    ImmutableRoaringBitmap[] potentialMatchDocIds = new ImmutableRoaringBitmap[potentialCoverH3Cells.size()];
+    int i = 0;
+    for (long h3IndexId : potentialCoverH3Cells) {
```

Comment:
Use iterator (`LongIterator.nextLong()`) instead to avoid the unnecessary boxing/unboxing, same for other places when iterating over `LongSet`

Reply:
sure. updated.